### PR TITLE
Add delay option and support for unknown options

### PR DESCRIPTION
**Pull request by kouhin**
We are using babel-watch in our project and found there are several problems.
I'm sorry that I make this all-in-one PR. 

Changes:
  - Prevent restart on compilation errors
  - Add `delay` option in order to delay restarting when multiple files changed at the same time (use lodash/debounce). 
  - <del>Support node harmony flag (especially useful when using node 7 async/await)</del>
  - Pass all unknown options to node process, e.g. `--debug`, `--harmony`, and even `--inspect` or `--debug-brk`

### CODE REVIEW

File: babel-watch.js
Comment (kmagiera) [inline babel-watch.js:73]: why is this needed?

File: package.json
Comment (kmagiera) [inline package.json]: I'd keep the list of dependencies short. Pulling in whole lodash just to have `debounce` method which can be implemented in 3 lines of code seems like a huge overhead to me. Would you be able to refactor this and drop dependency on lodash

File: babel-watch.js
Comment (kmagiera) [inline babel-watch.js:24]: I prefer to keep this option. Maybe you can use `-D` (capital D) for the delay?

**Review by kmagiera — CHANGES_REQUESTED** [review]
Thanks @kouhin for your PR! I've put a few comments on your changes. Would you be able to respond to them and I'll be happy to merge this in

File: babel-watch.js
Comment (kouhin) [reply babel-watch.js:73]: @kmagiera Unknown options will be allowed, otherwise it will cause an error.
e.g. `babel-watch --harmony --no-warnings index.js`

`--harmony` and `--no-warnings` are unknown options, they will be passed to node process. Just like `node --harmony --no-warnings index.js`

**Comment by kouhin** [comment]
@kmagiera BTW, this PR will support all the options including `--inspect` and `--debug-brk`. 
What do you think about reverting https://github.com/kmagiera/babel-watch/pull/30

**Comment by kouhin** [comment]
There is another bug in this PR. 
When I try `babel-watch --inspect=9222 index.js`, node process will be `node --inspect 9222 inspect`, and it causes an error that 9222 is not found. --- WIP

**Comment by kouhin** [comment]
@kmagiera I have fixed all the issues, and following commands work fine.

- `babel-watch --delay=3000 --harmony --debug=9222 --debug-brk -w ./src index.js`
- `babel-watch --delay=3000 --harmony --inspect=9222 -w ./src index.js`

File: babel-watch.js
Comment (kouhin) [reply babel-watch.js:24]: capital D is already used by `--disable-autowatch`, and node support `--debug` natively.
We can still use `babel-watch --debug`, and the option `--debug` will be passed to node.js.

**Comment by kouhin** [comment]
These features are already tested in our team for nearly 1 month and work fine.
The modified version can be installed by `npm i @kouhin/babal-watch --save-dev`

File: babel-watch.js
Comment (kmagiera) [inline babel-watch.js]: can you please use an arrow function instead to match the convention with the rest of the file

File: babel-watch.js
Comment (kmagiera) [inline babel-watch.js]: Why is the default delay set to 1sec? I prefer it to be 0 (it's quite a frequent case that you only work on a single file and you don't save it more frequently than few time in a minute, I'd prefer not to enforce people to wait an additional second for their app to restart by default)

File: babel-watch.js
Comment (kmagiera) [inline babel-watch.js:263]: Can't we just replace this whole block and just pass all the unknown options directly to `runnerExecArgv`? If there are options like `--debug 2388` there is no need in changing that to `--debug=2388`, right?

If so this would greatly simplified this fragment as we could just do:
```
runnerExecArgv = runnerExecArgv.concat(unknownOptions);
```

**Review by kmagiera — CHANGES_REQUESTED** [review]
Hey @kouhin! Sorry for getting back so late. Thanks for your work. I just have few last comments inline. Would you be able to reply to them and I can merge your change in. Thanks!

File: babel-watch.js
Comment (kouhin) [reply babel-watch.js]: Thank you.
I think maybe at least 100ms is necessary. Otherwise, when more than one files are generated at the same time, restarting will be horrible.
What about 100ms?

File: babel-watch.js
Comment (kouhin) [reply babel-watch.js]: OK, I will be rewrite it as an arrow function.

File: babel-watch.js
Comment (kouhin) [reply babel-watch.js:263]: This block is required, and I add it intentionally. 
When just using `runnerExecArgv = runnerExecArgv.concat(unknownOptions);`, and run 
`babel-watch --harmony --inspect -w ./src -D --debug=2388 ./src/server.js`

Args will be parsed to `[ '--harmony', '--inspect', '--debug', '2388' ]`,  and it causes an error: 
`Error: Cannot find module '/Users/kouhin/git/myproj/2388'`

**Comment by alloy** [comment]
The passing on of unknown options is something I really need. Is there anything specific blocking this from being merged?

If it’s around the detection of unknown options, another approach might be to use [a double dash](http://unix.stackexchange.com/questions/11376/what-does-double-dash-mean-also-known-as-bare-double-dash) to denote the end of `babel-watch`’s parameter list? E.g.

```
$ babel-watch -w ./src ./src/server.js -- --harmony --inspect -D --debug=2388
```

**Review by yormi — COMMENTED** [review]
`requiredFiles` on line 152 is not defined and crash `babel-watch` when not using `--disable-autowatch`

And I also think using the idea here https://github.com/component/debounce would be better than setTimeout to delay the restarts. It won't make you wait `delay` ms everytime but it'll just delay the restart call if it hasn't been `delay` ms.

File: babel-watch.js
Comment (yormi) [inline babel-watch.js]: `requiredFiles` not defined

**Comment by kouhin** [comment]
@yormi Thank you. Fixed! 

> And I also think using the idea here https://github.com/component/debounce would be better than setTimeout to delay the restarts. It won't make you wait delay ms everytime but it'll just delay the restart call if it hasn't been delay ms.

I think @kmagiera doesn't want to introduce extra library only for debounce.

**Comment by yormi** [comment]
Sweet ! Pretty quick :)

We can probably avoid introducing the dependency. It's barely 50 lines of code: https://github.com/component/debounce/blob/master/index.js.

Moreover, if I'm not mistaken, the `debounce` code would need to be adapted. Otherwise, on a pull of 10 files, the app would still restart 10 times but with the delay in between each time.

**Comment by coodoo** [comment]
`--harmony` is a much needed flag, anyway to speed this up and get it merged? :D

**Comment by kouhin** [comment]
@coodoo You can use this scoped package @kouhin/babel-watch until this PR is merged. 

`npm install @kouhin/babel-watch --save-dev`

@yormi 
> Moreover, if I'm not mistaken, the debounce code would need to be adapted. Otherwise, on a pull of 10 files, the app would still restart 10 times but with the delay in between each time.

Thank you! I'll confirm it later.

**Comment by kouhin** [comment]
@yormi I have check the action in my project. It seems OK. 
I inserted a console.info in `handleChange` to show changed files and another console.info in `restartAppImmediate()` to show 'restart'.
When I changed the branch, it showed there were 4 files changed at the same time, and only restarted for 1 time.

**Comment by yormi** [comment]
@kouhin Sorry for the late answer. University is keeping me pretty busy... ok, ok I admit, there's other stuff aside studying as well :P

Cool ! Seems to work as we want it to ! Have you added the debounce after all ?

On another note, without the `--disable-autowatch`, the app is not restarted more than once per file even after multiple modifs. It's not a big deal though since the `--watch` argument makes it work perfectly. Just dropping here if someone gets stuck with that !

Thanks for the good work :)

File: babel-watch.js
Comment (0x7f) [inline babel-watch.js:269]: Use space instead of `=`:
```
$ touch test.js
$ node --require=./test.js
node: bad option: --require=./test.js
$ node --require ./test.js
>
```

Comment (kouhin) [reply babel-watch.js:269]: It can't be so easy to solve this problem just by using space instead of `=`, I think.

Following POSIX docs https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

> To specify an argument for a long option, write ‘--name=value’. This syntax enables a long option to accept an argument that is itself optional.

node options:

```
❯ node --help
Usage: node [options] [ -e script | script.js ] [arguments]
       node debug script.js [arguments]

Options:
  -v, --version         print Node.js version
  -e, --eval script     evaluate script
  -p, --print           evaluate script and print result
  -c, --check           syntax check script without executing
  -i, --interactive     always enter the REPL even if stdin
                        does not appear to be a terminal
  -r, --require         module to preload (option can be repeated)
  --no-deprecation      silence deprecation warnings
  --trace-deprecation   show stack traces on deprecations
  --throw-deprecation   throw an exception anytime a deprecated function is used
  --no-warnings         silence all process warnings
  --trace-warnings      show stack traces on process warnings
  --trace-sync-io       show stack trace when use of sync IO
                        is detected after the first tick
  --track-heap-objects  track heap object allocations for heap snapshots
  --prof-process        process v8 profiler output generated
                        using --prof
  --zero-fill-buffers   automatically zero-fill all newly allocated
                        Buffer and SlowBuffer instances
  --v8-options          print v8 command line options
  --v8-pool-size=num    set v8's thread pool size
  --tls-cipher-list=val use an alternative default TLS cipher list
  --openssl-config=path load OpenSSL configuration file from the
                        specified path
  --icu-data-dir=dir    set ICU data load path to dir
                        (overrides NODE_ICU_DATA)
  --preserve-symlinks   preserve symbolic links when resolving
                        and caching modules.

Environment variables:
NODE_PATH                ':'-separated list of directories
                         prefixed to the module search path.
NODE_DISABLE_COLORS      set to 1 to disable colors in the REPL
NODE_ICU_DATA            data path for ICU (Intl object) data
NODE_REPL_HISTORY        path to the persistent REPL history file

Documentation can be found at https://nodejs.org/
```

I've no idea how to solve it. Maybe better to use other parser instead of Commander.

Comment (0x7f) [reply babel-watch.js:269]: You are right that posix args usually use `=` for arguments with values, but node.js is behaving really awkward here:
```
$ node --require ./test.js
>
$ node --require=./test.js
node: bad option: --require=./test.js
```
and:
```
$ node --v8-pool-size 10000000
node: bad option: --v8-pool-size
$ node --v8-pool-size=10000000
>
```
For parsing `--require` they explicitly address the next argument: https://github.com/nodejs/node/blob/291b599bba0f484f35a2deebb2350b178c1f1758/src/node.cc#L3673-L3680 but for the v8 argument, they require the equal sign: https://github.com/nodejs/node/blob/291b599bba0f484f35a2deebb2350b178c1f1758/src/node.cc#L3725

If you find an argument starting with `--` and containing a `=`, you just copy one. But if you find one with `--` but without `=`, it might be a flag (e.g. `--zero-fill-buffers`) or an argument with a value ( `--require ./test.js`). You could maintain a list with arguments that require values, but this is ugly and might change with future version...

So yeah, you are right, there is no easy solution to that. I have no idea what to do either :(

**Comment by kmagiera** [comment]
@vampirical would you have time to take a look at this PR? I'd appreciate your opinion and decision whether and in what shape this can be merged. Thanks in advance!

**Comment by STRML** [comment]
This will be rather difficult to merge now that delay is in master. The unknown options bits are decent but seem to introduce more problems than they fix. I recommend closing until @kouhin has the time to rewrite the options parsing on master.

**Comment by kmagiera** [comment]
I don't think it is worth @kouhin's time to rewrite until maintainers agree that we would like to merge this in. In the past I tried adding a few different people who volunteered to evaluate and review PRs but unfortunately most of the PRs still haven't been addressed. For that reason I'd prefer to keep this open and actually have a maintainer decide whether we don't want this functionality or ask the author to update it and give them some time prior to closing.

**Comment by STRML** [comment]
I'd be happy to review this if the conflicts are resolved. I just had a go at unknown options, I think it's better this way, but it's a bit complicated to deal with `program.args`, especially if additional args are passed to the script. For instance, how do we handle:

```
babel-watch --v8-pool-size=10 --require./foo.js app.js --some-app-switch
```

Handling this well is going to require writing a few short test cases IMO.

**Comment by STRML** [comment]
The module [string-argv](https://github.com/mccormicka/string-argv) seems to handle this well, and it seems `--require` was fixed in Node.

However, `process.argv` is a mess and `commander` does almost nothing to help us in the latest versions. I think we need to switch to yargs.

**Comment by STRML** [comment]
In `7.4.0` we now support `BABEL_WATCH_NODE_OPTIONS`, which essentially just passes `NODE_OPTIONS` along to the child. It should, therefore, support any node option you wish to support. I've kept `--debug`/`--debug-brk`, and `--inspect`/`--inspect-brk` as top-level options because they are convenient, but use the env for anything else.

**Pull request closed without merging**
